proposals/p2274.md

```diff
@@ -0,0 +1,384 @@
+have about it, so long as we have enough type information to establish that the
+expression is valid.
+
+The primary problem this proposal needs to solve is how to support both
+array-like and slice-like indexing semantics while satisfying those
+requirements.
+
+## Proposal
+
+Carbon will support indexing using the conventional `a[i]` subscript syntax.
+When `a` is an l-value, the result of subscripting will always be an l-value,
+but when `a` is an r-value, the result can be an l-value or an r-value,
+depending on which interface the type implements:
+
+-   If subscripting an r-value produces an r-value result, as with an array, the
+    type should implement `IndexWith`.
+-   If subscripting an r-value produces an l-value result, as with a slice, the
+    type should implement `IndirectIndexWith`.
+
+`IndirectIndexWith` is a subtype of `IndexWith`, and subscript expressions are
+rewritten to method calls on `IndirectIndexWith` if the type is known to
+implement that interface, or to method calls on `IndexWith` otherwise.
+
+A type can implement at most one of these two interfaces. `IndirectIndexWith`
+provides a final blanket `impl` of `IndexWith`, which ensures coherence.
+
+One implication of this design is that the requirement of coherence must apply
+to value categories as well as types (and for those purposes, "lvalue" is a
+subcategory of "rvalue"). This implies that for any valid expression containing
+an lvalue sub-expression, replacing that sub-expression with an rvalue
+expression cannot change the validity or semantics of the outer expression, so
+long as the rvalue expression has the same type and evaluates to the same value.
+This, in turn, ensures that once we know a value implements `IndexWith`,
+additionally learning that it implements `IndirectIndexWith` can't invalidate
+code that previously typechecked, or change its semantics.
+
+## Details
+
+A subscript expression has the form "_lhs_ `[` _index_ `]`". As in C++, this
+syntax has the same precedence as `.`, `->`, and function calls, and associates
+left-to-right with all of them.
+
+Its semantics are defined in terms of the following interfaces:
+
+```
+interface IndexWith(SubscriptType:! Type) {
+  let ElementType:! Type;
+  fn At[me: Self](subscript: SubscriptType) -> ElementType;
+  fn Addr[addr me: Self*](subscript: SubscriptType) -> ElementType*;
+}
+
+interface IndirectIndexWith(SubscriptType:! Type) {
+  impl as IndexWith(SubscriptType);
+  fn Addr[me: Self](subscript: SubscriptType) -> ElementType*;
+}
+```
+
+A subscript expression where _lhs_ has type `T` and _index_ has type `I` is
+rewritten based on the value category of _lhs_ and whether `T` implements
+`IndirectIndexWith(I)`:
+
+-   If `T` implements `IndirectIndexWith(I)`, the expression is rewritten to
+    "`*((` _lhs_ `).(IndirectIndexWith(I).Addr)(` _index_ `))`".
+-   Otherwise, if _lhs_ is an l-value, the expression is rewritten to "`*((`
+    _lhs_ `).(IndexWith(I).Addr)(` _index_ `))`".
```

Comment:
Is this guaranteed, or is an implementation permitted to call `At` instead if the result is used in a context that only requires an rvalue? Eg:
```
var a: Array(i32, 5);
// Is this required to call `Addr`, or would it be permissible to use `At` instead?
var b: i32 = a[4];
```

Reply:
I worry if the implementation does that, user code will try to leverage it to build `std::map`-style APIs where `Addr` can insert a new element. We might be able to avoid that risk if the implementation is sufficiently unreliable about when it uses `At` instead, but that seems like it would also undermine the benefits of allowing it to do that at all. So my inclination for now is to make this a guarantee.

Reply:
Seems reasonable to me. Is this worth capturing in an "Alternative considered"?

Reply:
I'm not sure. What would be the advantages of doing it that way?

Reply:
`At` may be a little more efficient in cases where both `At` and `Addr` work. It might also provide a better safety story, depending on where we end up with our value semantics model.

Reply:
OK, done.